src/NameService/NameService.cs

```diff
@@ -615,5 +606,53 @@ private static bool CheckIPv6(string ipv6)
             }
             return true;
         }
+        
+        /// <summary>
+        /// Checks provided name for validness and returns corresponding token ID.
+        /// </summary>
+        private static string tokenIDFromName(string name)
+        {
+            string[] fragments = SplitAndCheck(name, true);
+            if (fragments is null) throw new FormatException("The format of the name is incorrect.");
+            if (fragments.Length == 1) return name;
+            return name[^(fragments[^2].Length + fragments[^1].Length + 1)..];
+        }
+
+        /// <summary>
+        /// Retrieves NameState from storage and checks that it's not expired as far as the parent domain.
+        /// </summary>
+        private static NameState getNameState(StorageMap nameMap, string tokenId)
+        {
+            ByteString tokenBytes = nameMap[GetKey(tokenId)];
+            if (tokenBytes is null) throw new InvalidOperationException("Unknown token.");
+            NameState token = (NameState)StdLib.Deserialize(tokenBytes);
+            token.EnsureNotExpired();
+            string[] fragments = StdLib.StringSplit(tokenId, ".");
+            if (parentExpired(nameMap, 1, fragments)) throw new InvalidOperationException("Parent domain has expired.");
+            return token;
+        }
+
+        /// <summary>
+        /// Returns true if any domain from fragments doesn't exist or expired.
+        /// </summary>
+        /// <param name="nameMap">Registered domain names storage map.</param>
+        /// <param name="first">The deepest subdomain to check.</param>
+        /// <param name="fragments">The array of domain name fragments.</param>
+        /// <returns>Whether any domain fragment doesn't exist or expired.</returns>
+        private static bool parentExpired(StorageMap nameMap, int first, string[] fragments)
+        {
+            int last = fragments.Length - 1;
+            string name = fragments[last];
+            for (int i = last; i >= first; i--) {
+                if (i != last) {
+                    name = fragments[i] + "." + name;
+                }
+                ByteString buffer = nameMap[GetKey(name)];
+                if (buffer is null) return true;
+                NameState token = (NameState)StdLib.Deserialize(buffer);
+                if (Runtime.Time >= token.Expiration) return true;
```

Comment:
Is not better to store the min expiration time into `NameState`, seems expensive to me the method `getNameState`

Reply:
Then on every `Renew` we have to iterate over the whole set of dependant token children and re-calculate minimum expiration time for each `NameState` which implies retrieving the whole set of parent domains.

Reply:
I'd expect there to be ~2-3 iterations typically which doesn't seem to be a lot. Many of the affected methods are safe (OwnerOf, Properties, IsAvailable, etc.), so they'll be mostly invoked in test mode, not on-chain. Non-safe affected methods can have more problems with the alternative layout as with the `Renew` case above.